atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java

```diff
@@ -2513,38 +2571,89 @@ private void logCommitLockTenureExceeded(
+        long cellMetadataCount = 0L;
+        long rowMetadataCount = 0L;
         for (TableReference tableRef : writesByTable.keySet()) {
             ConflictHandler conflictHandler = getConflictHandlerForTable(tableRef);
+
+            int previousLockCount = lockDescriptors.size();
+            int previousMetadataCount = lockDescriptorToChangeMetadata.size();
             if (conflictHandler.lockCellsForConflicts()) {
-                NavigableSet<Cell> cellsToLock = getLocalWrites(tableRef).keySet();
-                for (Cell cell : cellsToLock) {
-                    result.add(AtlasCellLockDescriptor.of(
-                            tableRef.getQualifiedName(), cell.getRowName(), cell.getColumnName()));
-                }
-                cellLockCount = cellsToLock.size();
+                collectCellLocks(lockDescriptors, lockDescriptorToChangeMetadata, tableRef);
             }
+            cellLockCount += lockDescriptors.size() - previousLockCount;
+            cellMetadataCount += lockDescriptorToChangeMetadata.size() - previousMetadataCount;
 
+            previousLockCount = lockDescriptors.size();
+            previousMetadataCount = lockDescriptorToChangeMetadata.size();
             if (conflictHandler.lockRowsForConflicts()) {
-                Cell lastCell = null;
-                for (Cell cell : getLocalWrites(tableRef).keySet()) {
-                    if (lastCell == null || !Arrays.equals(lastCell.getRowName(), cell.getRowName())) {
-                        result.add(AtlasRowLockDescriptor.of(tableRef.getQualifiedName(), cell.getRowName()));
-                        rowLockCount++;
-                    }
-                    lastCell = cell;
-                }
+                collectRowLocks(lockDescriptors, lockDescriptorToChangeMetadata, tableRef);
             }
+            rowLockCount += lockDescriptors.size() - previousLockCount;
+            rowMetadataCount += lockDescriptorToChangeMetadata.size() - previousMetadataCount;
         }
-        result.add(AtlasRowLockDescriptor.of(
+        lockDescriptors.add(AtlasRowLockDescriptor.of(
                 TransactionConstants.TRANSACTION_TABLE.getQualifiedName(),
                 TransactionConstants.getValueForTimestamp(getStartTimestamp())));
         cellCommitLocksRequested = cellLockCount;
         rowCommitLocksRequested = rowLockCount + 1;
-        return result;
+        cellChangeMetadataSent = cellMetadataCount;
+        rowChangeMetadataSent = rowMetadataCount;
+        return LocksAndMetadata.of(
+                lockDescriptors,
+                // For now, lock request metadata only consists of change metadata. If it is absent, we can save
+                // computation by not doing an index encoding
+                lockDescriptorToChangeMetadata.isEmpty()
+                        ? Optional.empty()
+                        : Optional.of(LockRequestMetadata.of(lockDescriptorToChangeMetadata)));
+    }
+
+    private void collectCellLocks(
```

Comment:
I refactored the way we acquire commit locks a bit because the level of indentation was getting way too high when adding metadata